packages/@atjson/document/src/annotation.ts

```diff
@@ -1,18 +1,217 @@
+      this.end -= length;
+
+    } else {
+
+      if (change.end < this.end) {
+
+        // Annotation spans the whole deleted text, so just truncate the end of
+        // the annotation (shrink from the right).
+        //   [             ]
+        // ------*------*---------
+        if (change.start > this.start) {
+          this.end -= length;
+
+        // Annotation occurs within the deleted text, affecting both start and
+        // end of the annotation, but by only part of the deleted text length.
+        //         [         ]
+        // ---*---------*------------
+        } else if (change.start <= this.start) {
+          this.start -= this.start - change.start;
+          this.end -= length;
+        }
+
+      } else if (change.end >= this.end) {
+
+        //             [  ]
+        //          [     ]
+        //          [         ]
+        //              [     ]
+        //    ------*---------*--------
+        if (change.start <= this.start) {
+          this.start = change.start;
+          this.end = change.start;
+
+        //       [        ]
+        //    ------*---------*--------
+        } else if (change.start > this.start) {
+          this.end = change.start;
+        }
+      }
+    }
+  }
+
+  handleInsertion(change: Insertion) {
+    let length = change.text.length;
+
+    // The first two normal cases are self explanatory. Just adjust the annotation
+    // position, since there is never a case where we wouldn't want to.
+    if (change.start < this.start) {
+      this.start += length;
+      this.end += length;
+    } else if (change.start > this.start && change.start < this.end) {
+      this.end += length;
+
+    // In this case, however, the normal behaviour when inserting text at a
+    // point adjacent to an annotation is to drag along the end of the
+    // annotation, or push forward the beginning, i.e., the transform happens
+    // _inside_ an annotation to the left, or _outside_ an annotation to the right.
+    //
+    // Sometimes, the desire is to change the direction; this is provided below
+    // with the preserveAdjacentBoundaries switch.
+
+    // Default edge behaviour.
+    } else if (change.behaviour === AdjacentBoundaryBehaviour.default) {
+      if (change.start === this.start) {
+        this.start += length;
+        this.end += length;
+      } else if (change.start === this.end) {
+        this.end += length;
+      }
+
+    // Non-standard behaviour. Do nothing to the adjacent boundary!
+    } else if (change.behaviour === AdjacentBoundaryBehaviour.preserve && change.start === this.start) {
+      this.end += length;
 
-  attributes?: { [key: string]: any };
+    // no-op; we would delete the annotation, but we should defer to the
+    // annotation as to whether or not it's deletable, since some zero-length
+    // annotations should be retained.
+    // n.b. the += 0 is just to silence tslint ;-)
+    } else if (change.start === this.end)  {
+      this.end += 0;
+    }
```

Comment:
silencing tslint because it would have complained about an empty block? but do you need this block at all? or is it here just to serve as living documentation? (an `else` case would also do nothing)

Reply:
Living documentation here.